### PR TITLE
ElasticVectorSearch: Add in vector search backed by Elastic

### PR DESCRIPTION
**Pull request by sjwhitmore**
image

woo!

### CODE REVIEW

**Comment by sjwhitmore** [comment]
i can resolve test error here better after merging in my samantha/prompt_example_gen branch

**Review by hwchase17 — COMMENTED** [review]
can we follow the example in https://github.com/hwchase17/langchain/pull/68 of initializing the client in the init file? could also consider a classmethod implementation.... but i guess for now lets do it in init and refactor later if necc?

also will need to put imports there (dont think its nice to have them required for the whole package)

**Comment by sjwhitmore** [comment]
> can we follow the example in #68 of initializing the client in the init file? could also consider a classmethod implementation.... but i guess for now lets do it in init and refactor later if necc?
> 
> also will need to put imports there (dont think its nice to have them required for the whole package)

done!

File: langchain/elastic_vector_search.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+        Returns:
+            List of Documents most similar to the query.
+        """
+        embedding = self.embedding_function(query)
+        script_query = _default_script_query(embedding)
+        response = self.client.search(index=self.index_name, query=script_query)
+        i = 0
+        texts = []
+        for hit in response["hits"]["hits"]:
```
Comment (hwchase17) [inline langchain/elastic_vector_search.py:87]: could just do `response["hits"]["hits"][:k]` and not have to worry about the break or the `i` coutner?

**Review by hwchase17 — COMMENTED** [review]
could you add elasticsearch to the requirements.txt?

also could you add something quick to the readme (similar to other integratins) about (1) making sure to install the package, but also (2) a link to how to set up the backend?

File: langchain/elastic_vector_search.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+            )
+        self.embedding_function = embedding_function
+        self.index_name = index_name
+        try:
+            es_client = elasticsearch.Elasticsearch(elastic_url)  # noqa
+        except ValueError as e:
+            raise ValueError(
+                "Your elasticsearch client string is misformatted. " f"Got error: {e} "
+            )
+        self.client = elasticsearch.Elasticsearch(elastic_url)
```
Comment (hwchase17) [inline langchain/elastic_vector_search.py:69]: can we use es_client from above side we already initialized it?

File: langchain/elastic_vector_search.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            k: Number of Documents to return. Defaults to 4.
+
+        Returns:
+            List of Documents most similar to the query.
+        """
+        embedding = self.embedding_function(query)
+        script_query = _default_script_query(embedding)
+        response = self.client.search(index=self.index_name, query=script_query)
+        texts = []
+        for hit in response["hits"]["hits"][:k]:
```
Comment (hwchase17) [inline langchain/elastic_vector_search.py:86]: can probably be a one liner `texts = [hit["_source"]["text"] for hit in response["hits"]["hits"][:k]]`

**Review by hwchase17 — APPROVED** [review]
some final nits but otherwise lgtm!

File: README.md [hunk 4]
```diff
@@ -47,6 +47,11 @@ The following use cases require specific installs and api keys:
   - Install requirements with `pip install playwright`
 - _Wikipedia_:
   - Install requirements with `pip install wikipedia`
+- _Elasticsearch_:
+  - Install requirements with `pip install elasticsearch`
+  - Set up Elasticsearch backend. If you want to do locally, [this](https://www.elastic.co/guide/en/elasticsearch/reference/7.17/getting-started.html) is a good guide.
```
Comment (hwchase17) [inline README.md:52]: fyi added this @sjwhitmore

**Pull request merged**